components/places/src/storage.rs

```diff
@@ -263,21 +270,314 @@ fn add_visit(db: &impl ConnExt,
     Ok(RowId(rid))
 }
 
-// Currently not used - we update the frecency as we update the page info.
-pub fn update_frecency(db: &mut PlacesDb, id: RowId, redirect: Option<bool>) -> Result<()> {
-    let score = frecency::calculate_frecency(db.conn(),
-        &frecency::DEFAULT_FRECENCY_SETTINGS,
-        id.0, // TODO: calculate_frecency should take a RowId here.
-        redirect)?;
+// Support for Sync - in its own module to try and keep a delineation
+pub mod history_sync {
+    use super::*;
+    use std::collections::HashMap;
+    use history_sync::record::{HistoryRecord, HistoryRecordVisit};
+    use history_sync::{HISTORY_TTL};
+
+    #[derive(Debug)]
+    pub struct FetchedVisit {
+        pub is_local: bool,
+        pub visit_date: Timestamp,
+        pub visit_type: Option<VisitTransition>,
+    }
 
-    db.execute_named("
-        UPDATE moz_places
-        SET frecency = :frecency
-        WHERE id = :page_id",
-        &[(":frecency", &score), (":page_id", &id.0)])?;
+    impl FetchedVisit {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                is_local: row.get_checked("is_local")?,
+                visit_date: row.get_checked::<_, Option<Timestamp>>("visit_date")?
+                     .unwrap_or_default(),
+                visit_type: VisitTransition::from_primitive(
+                                row.get_checked::<_, Option<u8>>("visit_type")?.unwrap_or(0)
+                            ),
+            })
+        }
+    }
 
-    Ok(())
-}
+    #[derive(Debug)]
+    pub struct FetchedVisitPage {
+        pub url: Url,
+        pub guid: SyncGuid,
+        pub row_id: RowId,
+        pub title: String,
+    }
+
+    impl FetchedVisitPage {
+        pub fn from_row(row: &Row) -> Result<Self> {
+            Ok(Self {
+                url: Url::parse(&row.get_checked::<_, Option<String>>("url")?.expect("non null column"))?,
```

Comment:
I don't think theres ever really any reason to do get_checked for an Option<String> that we immediately do unwrap/expect on. This will just make our errors harder to handle and debug, since we'll panic instead of returning an error. `row.get_checked::<_, String>("url")?` will do what you want.

This applies below (and anywhere else) too.